xt_training/utils/logging.py

```diff
@@ -37,4 +44,70 @@ def flush(self):
         self.stdout.flush()
 
     def __del__(self):
-        self.close()
\ No newline at end of file
+        self.close()
+
+
+def _save_config(save_dir, config_path):
+    # Save config file
+    try:
+        if isinstance(config_path, str):
+            shutil.copy(config_path, os.path.join(save_dir, 'config.py'))
+        else:
+            shutil.copy(config_path['__file__'], os.path.join(save_dir, 'config.py'))
+    except shutil.SameFileError:
+        pass
+
+
+PATCH_HEADER = '''
+To replicate the git state used for this checkpoint, run the following:
+
+    $ git checkout <commit hash shown below>
+    $ git apply <path to this file>
+'''
+
+SESSION_HEADER = '''
+"""The following code represents the most recent 200 lines of the python history leading up to calling
+xt-training from an interactive python session. This is recorded to allow for reproduction of
+experiments. However, this is only possible if 200 lines (or less) of python history are sufficient
+to describe the state at runtime. For cleaner state logging, try using an IPython console instead 
+(ipython or jupyter)."""
+
+'''
+
+
+def _save_state(save_dir):
+    # If we are in a git repo, save git state file
+    try:
+        repo = git.Repo('.')
+        commit = repo.head.object.hexsha
+        untracked = repo.untracked_files
+        diff = repo.git.diff()
+
+        with open(os.path.join(save_dir, 'git.patch'), 'w') as f:
+            f.write(PATCH_HEADER)
+            f.write(f'\n\nCommit: {commit}')
+            f.write('\n\nUntracked files:\n')
+            f.write('\n'.join(untracked))
+            f.write('\n\n')
+            f.write(diff)
+    # Silently skip if no git repo is found
+    except:
+        pass
+
+    # Save session history (for interactive sessions only)
+    if hasattr(sys, 'ps1'):
+        # We are running in an interactive session
+        try:
+            if hasattr(main, 'In'):
+                with open(os.path.join(save_dir, 'session-ipython.py'), 'w') as f:
+                    f.write('\n'.join(main.In))
+            else:
+                with open(os.path.join(save_dir, 'session.py'), 'w') as f:
```

Comment:
Could we maybe just add a comment here that this if/else statement is distinguishing between an IPython session and a session using the built in python console? Just for future reference.